File: cdn-system/api/services/stats_range.go
package services

import (
	"strings"
	"time"
)

const statsTimeLayout = "2006-01-02 15:04:05"

// StatsRange defines a time window and bucket size for stats queries.
type StatsRange struct {
	Start       time.Time
	End         time.Time
	Bucket      time.Duration
	LabelFormat string
}

func ResolveStatsRange(rangeKey, startRaw, endRaw string, now time.Time) StatsRange {
	key := strings.ToLower(strings.TrimSpace(rangeKey))
	if key == "" {
		key = "30min"
	}
	switch key {
	case "today":
		start := beginningOfDay(now)
		return StatsRange{Start: start, End: now, Bucket: time.Hour, LabelFormat: "15:00"}
	case "yesterday":
		start := beginningOfDay(now).AddDate(0, 0, -1)
		end := endOfDay(start)
		return StatsRange{Start: start, End: end, Bucket: time.Hour, LabelFormat: "15:00"}
	case "7d", "7days", "7-day", "7":
		start := beginningOfDay(now).AddDate(0, 0, -6)
		return StatsRange{Start: start, End: now, Bucket: 24 * time.Hour, LabelFormat: "01-02"}
	case "30d", "30days", "30-day", "30":
		start := beginningOfDay(now).AddDate(0, 0, -29)
		return StatsRange{Start: start, End: now, Bucket: 24 * time.Hour, LabelFormat: "01-02"}
	case "last_month":
		start := beginningOfMonth(now).AddDate(0, -1, 0)
		end := endOfMonth(start)
		return StatsRange{Start: start, End: end, Bucket: 24 * time.Hour, LabelFormat: "01-02"}
	case "10min":
		return StatsRange{Start: now.Add(-10 * time.Minute), End: now, Bucket: time.Minute, LabelFormat: "15:04"}
	case "1h":
		return StatsRange{Start: now.Add(-1 * time.Hour), End: now, Bucket: time.Minute, LabelFormat: "15:04"}
	case "custom":
		if start, end, ok := parseCustomRange(startRaw, endRaw, now.Location()); ok {
			return buildCustomRange(start, end)
		}
	}
	return StatsRange{Start: now.Add(-30 * time.Minute), End: now, Bucket: time.Minute, LabelFormat: "15:04"}
}

func parseCustomRange(startRaw, endRaw string, loc *time.Location) (time.Time, time.Time, bool) {
	startRaw = strings.TrimSpace(startRaw)
	endRaw = strings.TrimSpace(endRaw)
	if startRaw == "" || endRaw == "" {
		return time.Time{}, time.Time{}, false
	}
	start, err1 := time.ParseInLocation(statsTimeLayout, startRaw, loc)
	end, err2 := time.ParseInLocation(statsTimeLayout, endRaw, loc)
	if err1 != nil || err2 != nil || end.Before(start) {
		return time.Time{}, time.Time{}, false
	}
	return start, end, true
}

func buildCustomRange(start, end time.Time) StatsRange {
	duration := end.Sub(start)
	if duration <= time.Hour {
		return StatsRange{Start: start, End: end, Bucket: time.Minute, LabelFormat: "15:04"}
	}
	if duration <= 24*time.Hour {
		return StatsRange{Start: start, End: end, Bucket: time.Hour, LabelFormat: "15:00"}
	}
	return StatsRange{Start: start, End: end, Bucket: 24 * time.Hour, LabelFormat: "01-02"}
}

func AlignToBucket(ts time.Time, bucket time.Duration) time.Time {
	if bucket >= 24*time.Hour {
		return beginningOfDay(ts)
	}
	return ts.Truncate(bucket)
}

func beginningOfDay(ts time.Time) time.Time {
	return time.Date(ts.Year(), ts.Month(), ts.Day(), 0, 0, 0, 0, ts.Location())
}

func beginningOfMonth(ts time.Time) time.Time {
	return time.Date(ts.Year(), ts.Month(), 1, 0, 0, 0, 0, ts.Location())
}

func endOfDay(ts time.Time) time.Time {
	return time.Date(ts.Year(), ts.Month(), ts.Day(), 23, 59, 59, 0, ts.Location())
}

func endOfMonth(ts time.Time) time.Time {
	startOfMonth := beginningOfMonth(ts)
	startNext := startOfMonth.AddDate(0, 1, 0)
	return startNext.Add(-time.Second)
}

File: cnn.net/src/Cnn.Api/Services/Stats/StatsRange.cs
namespace Cnn.Api.Services.Stats;

public sealed record StatsRange(DateTime Start, DateTime End, TimeSpan Bucket, string LabelFormat);

public static class StatsRangeResolver
{
    private const string StatsTimeLayout = "yyyy-MM-dd HH:mm:ss";

    public static StatsRange Resolve(string? rangeKey, string? startRaw, string? endRaw, DateTime now)
    {
        var key = (rangeKey ?? string.Empty).Trim().ToLowerInvariant();
        if (string.IsNullOrWhiteSpace(key))
        {
            key = "30min";
        }

        switch (key)
        {
            case "today":
                {
                    var start = BeginningOfDay(now);
                    return new StatsRange(start, now, TimeSpan.FromHours(1), "HH:00");
                }
            case "yesterday":
                {
                    var start = BeginningOfDay(now).AddDays(-1);
                    var end = EndOfDay(start);
                    return new StatsRange(start, end, TimeSpan.FromHours(1), "HH:00");
                }
            case "7d":
            case "7days":
            case "7-day":
            case "7":
                {
                    var start = BeginningOfDay(now).AddDays(-6);
                    return new StatsRange(start, now, TimeSpan.FromDays(1), "MM-dd");
                }
            case "30d":
            case "30days":
            case "30-day":
            case "30":
                {
                    var start = BeginningOfDay(now).AddDays(-29);
                    return new StatsRange(start, now, TimeSpan.FromDays(1), "MM-dd");
                }
            case "last_month":
                {
                    var start = BeginningOfMonth(now).AddMonths(-1);
                    var end = EndOfMonth(start);
                    return new StatsRange(start, end, TimeSpan.FromDays(1), "MM-dd");
                }
            case "10min":
                return new StatsRange(now.AddMinutes(-10), now, TimeSpan.FromMinutes(1), "HH:mm");
            case "1h":
                return new StatsRange(now.AddHours(-1), now, TimeSpan.FromMinutes(1), "HH:mm");
            case "custom":
                if (TryParseCustomRange(startRaw, endRaw, now, out var customStart, out var customEnd))
                {
                    return BuildCustomRange(customStart, customEnd);
                }
                break;
        }

        return new StatsRange(now.AddMinutes(-30), now, TimeSpan.FromMinutes(1), "HH:mm");
    }

    public static DateTime AlignToBucket(DateTime value, TimeSpan bucket)
    {
        if (bucket >= TimeSpan.FromDays(1))
        {
            return BeginningOfDay(value);
        }

        if (bucket == TimeSpan.Zero)
        {
            return value;
        }

        var ticks = value.Ticks / bucket.Ticks * bucket.Ticks;
        return new DateTime(ticks, value.Kind);
    }

    private static bool TryParseCustomRange(string? startRaw, string? endRaw, DateTime now, out DateTime start, out DateTime end)
    {
        start = DateTime.MinValue;
        end = DateTime.MinValue;
        startRaw = startRaw?.Trim();
        endRaw = endRaw?.Trim();
        if (string.IsNullOrWhiteSpace(startRaw) || string.IsNullOrWhiteSpace(endRaw))
        {
            return false;
        }

        if (!DateTime.TryParseExact(startRaw, StatsTimeLayout, null, System.Globalization.DateTimeStyles.None, out start))
        {
            return false;
        }

        if (!DateTime.TryParseExact(endRaw, StatsTimeLayout, null, System.Globalization.DateTimeStyles.None, out end))
        {
            return false;
        }

        if (end < start)
        {
            return false;
        }

        return true;
    }

    private static StatsRange BuildCustomRange(DateTime start, DateTime end)
    {
        var duration = end - start;
        if (duration <= TimeSpan.FromHours(1))
        {
            return new StatsRange(start, end, TimeSpan.FromMinutes(1), "HH:mm");
        }

        if (duration <= TimeSpan.FromDays(1))
        {
            return new StatsRange(start, end, TimeSpan.FromHours(1), "HH:00");
        }

        return new StatsRange(start, end, TimeSpan.FromDays(1), "MM-dd");
    }

    private static DateTime BeginningOfDay(DateTime value)
    {
        return new DateTime(value.Year, value.Month, value.Day, 0, 0, 0, value.Kind);
    }

    private static DateTime EndOfDay(DateTime value)
    {
        return new DateTime(value.Year, value.Month, value.Day, 23, 59, 59, value.Kind);
    }

    private static DateTime BeginningOfMonth(DateTime value)
    {
        return new DateTime(value.Year, value.Month, 1, 0, 0, 0, value.Kind);
    }

    private static DateTime EndOfMonth(DateTime value)
    {
        var start = BeginningOfMonth(value);
        var next = start.AddMonths(1);
        return next.AddSeconds(-1);
    }
}
